client/starwhale/core/dataset/copy.py

```diff
@@ -1,22 +1,90 @@
+import os
+from typing import List, Iterator, Optional
 from pathlib import Path
 
 from rich.progress import Progress
 
-from starwhale.utils import console
-from starwhale.consts import STANDALONE_INSTANCE
-from starwhale.base.bundle_copy import BundleCopy
+from starwhale.utils import console, load_yaml, NoSupportError
+from starwhale.consts import (
+    STANDALONE_INSTANCE,
+    DEFAULT_MANIFEST_NAME,
+    ARCHIVED_SWDS_META_FNAME,
+)
+from starwhale.utils.fs import ensure_dir
+from starwhale.base.bundle_copy import FileDesc, FileType, BundleCopy
 
+from .store import DatasetStorage
 from .tabular import TabularDataset
 
 
 class DatasetCopy(BundleCopy):
+    def upload_files(self, workdir: Path) -> Iterator[FileDesc]:
+        _manifest = load_yaml(workdir / DEFAULT_MANIFEST_NAME)
+        for _k in _manifest["signature"]:
+            _size, _, _hash = _k.split(":")
+            # TODO: head object by hash name at first
+            _path = workdir / "data" / _hash[: DatasetStorage.short_sign_cnt]
+            yield FileDesc(
+                path=_path,
+                name=os.path.basename(_path),
+                size=_size,
+                file_type=FileType.DATA,
+                signature=_hash,
+            )
+
+        _meta_names = [ARCHIVED_SWDS_META_FNAME]
+
+        for _n in _meta_names:
+            _path = workdir / _n
+
+            yield FileDesc(
+                path=_path,
+                name=os.path.basename(_path),
+                size=_path.stat().st_size,
+                file_type=FileType.SRC_TAR,
+                signature="",
+            )
+
+    def download_files(self, workdir: Path) -> Iterator[FileDesc]:
+        ensure_dir(workdir / "data")
+        _manifest = load_yaml(workdir / DEFAULT_MANIFEST_NAME)
+        for _k in _manifest.get("signature", []):
+            # TODO: parallel download
+            _size, _algo, _hash = _k.split(":")
+            if _algo != DatasetStorage.object_hash_algo:
+                raise NoSupportError(f"download file hash algorithm {_algo}")
+
+            _dest = DatasetStorage._get_object_store_path(_hash)
+
+            yield FileDesc(
+                path=_dest,
+                signature=_hash,
+                size=_size,
+                name=_hash[: DatasetStorage.short_sign_cnt],
+                file_type=FileType.DATA,
+                existed=False if (not _dest.exists() or self.force) else True,
```

Comment:
`existed=False if (not _dest.exists() or self.force) else True` -> `_dest.exists() and not force` ?

Reply:
done